### PR TITLE
Avoid reading _metadata on every worker

### LINKED ISSUE #5842: `_metadata` is re-read for every partition when using `fastparquet` engine
## Background
Given a (poorly) partitioned dataset of 15000 partitions of ~3k rows each, with a `_metadata` file, we evaluate e.g.

```python
df = dd.read_parquet(source, engine=engine)  # L1: mostly read _metadata
df.head()  # L2: single task
df.size.compute()  # L3: 15000 tasks
```

with engine set to either `fastparquet` or `pyarrow`. After [this speed-up](https://github.com/dask/dask/pull/5821), L1 takes ~25s with `fastparquet` vs 215s with `pyarrow`.

However, L2 takes ~20s with `fastparquet` vs ~200-300ms with `pyarrow`.

Similarly, L3 takes ~20s **per partition** (`read-parquet-...` task) on the workers - this appears to be largely due to repeated re-reading the metadata file, as confirmed by removing the `_metadata` file after executing L1.

This makes processing the data almost 100x slower than it could be, and leads to `O(n^2)` time in size of data when partitions are small.

### PR DESCRIPTION
**Pull request by rjzamora**
~Closes~ Partially addresses #5842 

For `FastParquetEngine`,  we are currently re-reading the "_metadata" file for every partition in many cases. This PR will avoid doing so whenever possible.

- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by rjzamora** [comment]
@ig248 - Note that this PR addresses some of [your suggestions](https://github.com/dask/dask/issues/5842#issuecomment-590415160) from #5842 (although the current changes seem most helpful for non-partitioned datasets, since the `"_metadata"` file should no longer be re-read at all).

**Comment by mrocklin** [comment]
cc @martindurant if you have a moment to review

**Comment by martindurant** [comment]
Is this still WIP, or are you ready for me to have a look, @rjzamora ?

**Comment by rjzamora** [comment]
Thanks @martindurant!  I am not planning to add anything here today, so a review at your leisure is certainly appreciated :)

File: dask/dataframe/io/parquet/fastparquet.py [hunk 1]
```diff
@@ -344,23 +347,58 @@ def read_metadata(
         # Create `parts`
         # This is a list of row-group-descriptor dicts, or file-paths
         # if we have a list of files and gather_statistics=False
+        base_path = (base_path or "") + fs.sep
```
Comment (martindurant) [inline dask/dataframe/io/parquet/fastparquet.py]: If we're doing direct path manipulations, we should maybe start a HTTP server and test against it

File: dask/dataframe/io/parquet/fastparquet.py [hunk 2]
```diff
@@ -344,23 +347,58 @@ def read_metadata(
...
             partsin = pf.row_groups
-            if fast_metadata:
-                pf = (paths, gather_statistics)
+            partitions = pf.info.get("partitions", None)
+            if partitions and not fast_metadata:
+                # We have partitions, and do not have
+                # a "_metadata" file for the worker to read.
+                # Therefore, we need to pass the pf object in
+                # the task graph
+                pf_deps = pf
```
Comment (martindurant) [inline dask/dataframe/io/parquet/fastparquet.py]: There was some code to strip down the pf instance to avoid thrift serialisation costs. Is that still happening?

File: dask/dataframe/io/parquet/fastparquet.py [hunk 3]
```diff
@@ -344,23 +347,58 @@ def read_metadata(
...
+            pf_deps = None
             partsin = parts
-            pf = None
         parts = []
+        i_path = 0
+        path_last = None
         for i, piece in enumerate(partsin):
-            if pf and not fast_metadata:
-                for col in piece.columns:
-                    col.meta_data.statistics = None
```
Comment (martindurant) [inline dask/dataframe/io/parquet/fastparquet.py]: I think this is what I was referring to above, making the pf instance smaller

File: dask/dataframe/io/parquet/fastparquet.py [hunk 4]
```diff
@@ -344,23 +347,58 @@ def read_metadata(
...
-                    col.meta_data.encoding_stats = None
-            piece_item = i if pf else piece
+            if partitions and fast_metadata:
```
Comment (martindurant) [inline dask/dataframe/io/parquet/fastparquet.py]: The three conditions here are a bit hard to parse. Of course, the situation **is** complicated. Perhaps up front, we should enumerate the cases and label the branches as required:
- a single file
- a directory of files with a _metadata
- a directory of files without _metadata
  - stats requested
  - stats unnecessary

File: dask/dataframe/io/parquet/fastparquet.py (hunk 3 above)
Comment (rjzamora) [reply dask/dataframe/io/parquet/fastparquet.py]: Ah - Right!  That should certainly stay in.

**Comment by martindurant** [comment]
Failure is a RuntimeWarning on py38 in test_cov (array?) - perhaps a new compiler version of numpy? In other words, unrelated to this PR.

**Comment by rjzamora** [comment]
> Glad to see it passing! Do we have benchmarks?

Just a note: I will revisit this soonish, but I did not see significant performance improvements from these changes when I last checked.  For the partitioned dataset case, we no longer spend much time in `_determine_pf_parts`, but we still spend a lot of time creating `pf` for each partition.

**Comment by martindurant** [comment]
Can you please merge from master? I believe things should pass now.

**Comment by TomAugspurger** [comment]
@martindurant did you want to give this another look, or was https://github.com/dask/dask/pull/6017#issuecomment-610564394 saying this was good?
